File: internal/game/game.go
package game

import (
	"errors"
	"math"
	"sort"
)

type Side string

const (
	Allies Side = "allies"
	Axis   Side = "axis"
)

func (s Side) Opponent() Side {
	if s == Allies {
		return Axis
	}
	return Allies
}

type Phase string

const (
	Waiting  Phase = "waiting"
	Playing  Phase = "playing"
	Finished Phase = "finished"
)

type UnitType string

const (
	Rifleman UnitType = "rifleman"
	Assault  UnitType = "assault"
	Gunner   UnitType = "gunner"
	Tank     UnitType = "tank"
)

type UnitSpec struct {
	Name        string  `json:"name"`
	Cost        float64 `json:"cost"`
	MaxHP       float64 `json:"maxHp"`
	Speed       float64 `json:"speed"`
	Range       float64 `json:"range"`
	Damage      float64 `json:"damage"`
	Cooldown    float64 `json:"cooldown"`
	Radius      float64 `json:"radius"`
	Description string  `json:"description"`
}

var Catalog = map[UnitType]UnitSpec{
	Rifleman: {
		Name: "Rifleman", Cost: 60, MaxHP: 100, Speed: 58, Range: 175,
		Damage: 17, Cooldown: 0.9, Radius: 13,
		Description: "Infanteri serbaguna dengan jarak tembak menengah.",
	},
	Assault: {
		Name: "Assault", Cost: 90, MaxHP: 135, Speed: 82, Range: 78,
		Damage: 31, Cooldown: 0.68, Radius: 14,
		Description: "Pasukan cepat untuk menembus garis dan parit lawan.",
	},
	Gunner: {
		Name: "Gunner", Cost: 135, MaxHP: 165, Speed: 39, Range: 245,
		Damage: 12, Cooldown: 0.27, Radius: 16,
		Description: "Penekan jarak jauh dengan laju tembak tinggi.",
	},
	Tank: {
		Name: "Tank", Cost: 320, MaxHP: 590, Speed: 31, Range: 275,
		Damage: 62, Cooldown: 1.45, Radius: 26,
		Description: "Unit berat yang mahal, lambat, dan sangat tahan.",
	},
}

type PlayerState struct {
	Resources float64 `json:"resources"`
	BaseHP    float64 `json:"baseHp"`
	MaxBaseHP float64 `json:"maxBaseHp"`
}

type Unit struct {
	ID          int64    `json:"id"`
	Type        UnitType `json:"type"`
	Side        Side     `json:"side"`
	X           float64  `json:"x"`
	Y           float64  `json:"y"`
	HP          float64  `json:"hp"`
	MaxHP       float64  `json:"maxHp"`
	Radius      float64  `json:"radius"`
	AttackPulse int64    `json:"attackPulse"`
	cooldown    float64
}

type Trench struct {
	X     float64 `json:"x"`
	Width float64 `json:"width"`
}

type Snapshot struct {
	Phase    Phase                `json:"phase"`
	Winner   Side                 `json:"winner,omitempty"`
	Reason   string               `json:"reason,omitempty"`
	Elapsed  float64              `json:"elapsed"`
	Tick     int64                `json:"tick"`
	Width    float64              `json:"width"`
	Height   float64              `json:"height"`
	Players  map[Side]PlayerState `json:"players"`
	Units    []Unit               `json:"units"`
	Trenches []Trench             `json:"trenches"`
}

type player struct {
	Resources float64
	BaseHP    float64
}

type Game struct {
	phase    Phase
	winner   Side
	reason   string
	elapsed  float64
	tick     int64
	nextID   int64
	width    float64
	height   float64
	allies   player
	axis     player
	units    []Unit
	trenches []Trench
}

const (
	startResources  = 220.0
	resourceRate    = 24.0
	resourceCap     = 999.0
	maxBaseHP       = 2200.0
	maxUnitsPerSide = 120
)

var (
	ErrNotPlaying        = errors.New("match is not currently playing")
	ErrUnknownUnit       = errors.New("unknown unit type")
	ErrInsufficientFunds = errors.New("insufficient resources")
	ErrInvalidSide       = errors.New("invalid side")
	ErrUnitCap           = errors.New("unit cap reached")
)

func New() *Game {
	return &Game{
		phase:  Waiting,
		nextID: 1,
		width:  2000,
		height: 720,
		allies: player{Resources: startResources, BaseHP: maxBaseHP},
		axis:   player{Resources: startResources, BaseHP: maxBaseHP},
		trenches: []Trench{
			{X: 430, Width: 150},
			{X: 810, Width: 130},
			{X: 1190, Width: 130},
			{X: 1570, Width: 150},
		},
	}
}

func (g *Game) Start() {
	if g.phase == Waiting {
		g.phase = Playing
	}
}

func (g *Game) Phase() Phase { return g.phase }

func (g *Game) Forfeit(loser Side, reason string) {
	if g.phase != Playing || (loser != Allies && loser != Axis) {
		return
	}
	g.phase = Finished
	g.winner = loser.Opponent()
	g.reason = reason
}

func (g *Game) Spawn(side Side, unitType UnitType) error {
	if g.phase != Playing {
		return ErrNotPlaying
	}
	if side != Allies && side != Axis {
		return ErrInvalidSide
	}
	spec, ok := Catalog[unitType]
	if !ok {
		return ErrUnknownUnit
	}
	if g.unitCount(side) >= maxUnitsPerSide {
		return ErrUnitCap
	}

	p := g.playerFor(side)
	if p.Resources < spec.Cost {
		return ErrInsufficientFunds
	}
	p.Resources -= spec.Cost

	x := 125.0
	if side == Axis {
		x = g.width - 125
	}

	lane := float64((g.nextID%4)-1) * 21
	g.units = append(g.units, Unit{
		ID:     g.nextID,
		Type:   unitType,
		Side:   side,
		X:      x,
		Y:      g.height*0.67 + lane,
		HP:     spec.MaxHP,
		MaxHP:  spec.MaxHP,
		Radius: spec.Radius,
	})
	g.nextID++
	return nil
}

func (g *Game) Step(dt float64) {
	if g.phase != Playing || dt <= 0 {
		return
	}
	if dt > 0.25 {
		dt = 0.25
	}

	g.elapsed += dt
	g.tick++
	g.allies.Resources = math.Min(resourceCap, g.allies.Resources+resourceRate*dt)
	g.axis.Resources = math.Min(resourceCap, g.axis.Resources+resourceRate*dt)

	for i := range g.units {
		if g.units[i].HP <= 0 {
			continue
		}
		if g.units[i].cooldown > 0 {
			g.units[i].cooldown -= dt
		}
		g.stepUnit(i, dt)
	}

	alive := g.units[:0]
	for _, unit := range g.units {
		if unit.HP > 0 {
			alive = append(alive, unit)
		}
	}
	g.units = alive

	sort.SliceStable(g.units, func(i, j int) bool {
		if g.units[i].X == g.units[j].X {
			return g.units[i].ID < g.units[j].ID
		}
		return g.units[i].X < g.units[j].X
	})

	g.resolveWinner()
}

func (g *Game) stepUnit(index int, dt float64) {
	unit := &g.units[index]
	spec := Catalog[unit.Type]

	targetIndex, distance := g.nearestEnemy(index)
	if targetIndex >= 0 && distance <= spec.Range+g.units[targetIndex].Radius {
		if unit.cooldown <= 0 {
			damage := spec.Damage * g.damageMultiplierAt(g.units[targetIndex].X, g.units[targetIndex].Type)
			g.units[targetIndex].HP -= damage
			unit.cooldown = spec.Cooldown
			unit.AttackPulse++
		}
		return
	}

	baseX := 72.0
	if unit.Side == Allies {
		baseX = g.width - 72
	}
	baseDistance := math.Abs(baseX - unit.X)
	if baseDistance <= spec.Range+42 {
		if unit.cooldown <= 0 {
			target := g.playerFor(unit.Side.Opponent())
			target.BaseHP -= spec.Damage
			unit.cooldown = spec.Cooldown
			unit.AttackPulse++
		}
		return
	}

	if g.friendlyBlocked(index) {
		return
	}

	direction := 1.0
	if unit.Side == Axis {
		direction = -1
	}
	unit.X += direction * spec.Speed * dt
	unit.X = math.Max(85, math.Min(g.width-85, unit.X))
}

func (g *Game) nearestEnemy(index int) (int, float64) {
	unit := g.units[index]
	bestIndex := -1
	bestDistance := math.MaxFloat64
	for i := range g.units {
		if i == index || g.units[i].HP <= 0 || g.units[i].Side == unit.Side {
			continue
		}
		distance := math.Abs(g.units[i].X - unit.X)
		if distance < bestDistance {
			bestDistance = distance
			bestIndex = i
		}
	}
	return bestIndex, bestDistance
}

func (g *Game) friendlyBlocked(index int) bool {
	unit := g.units[index]
	for i := range g.units {
		if i == index || g.units[i].HP <= 0 || g.units[i].Side != unit.Side {
			continue
		}
		delta := g.units[i].X - unit.X
		if unit.Side == Axis {
			delta = -delta
		}
		if delta > 0 && delta < unit.Radius+g.units[i].Radius+9 {
			return true
		}
	}
	return false
}

func (g *Game) damageMultiplierAt(x float64, targetType UnitType) float64 {
	if targetType == Tank {
		return 1
	}
	for _, trench := range g.trenches {
		if math.Abs(x-trench.X) <= trench.Width/2 {
			return 0.68
		}
	}
	return 1
}

func (g *Game) resolveWinner() {
	if g.allies.BaseHP > 0 && g.axis.BaseHP > 0 {
		return
	}
	g.phase = Finished
	switch {
	case g.allies.BaseHP <= 0 && g.axis.BaseHP <= 0:
		g.winner = "draw"
		g.reason = "both headquarters were destroyed"
	case g.axis.BaseHP <= 0:
		g.winner = Allies
		g.reason = "axis headquarters destroyed"
	default:
		g.winner = Axis
		g.reason = "allied headquarters destroyed"
	}
}

func (g *Game) unitCount(side Side) int {
	count := 0
	for i := range g.units {
		if g.units[i].HP > 0 && g.units[i].Side == side {
			count++
		}
	}
	return count
}

func (g *Game) playerFor(side Side) *player {
	if side == Allies {
		return &g.allies
	}
	return &g.axis
}

func (g *Game) Snapshot() Snapshot {
	units := make([]Unit, len(g.units))
	copy(units, g.units)
	trenches := make([]Trench, len(g.trenches))
	copy(trenches, g.trenches)
	return Snapshot{
		Phase:   g.phase,
		Winner:  g.winner,
		Reason:  g.reason,
		Elapsed: g.elapsed,
		Tick:    g.tick,
		Width:   g.width,
		Height:  g.height,
		Players: map[Side]PlayerState{
			Allies: {Resources: g.allies.Resources, BaseHP: math.Max(0, g.allies.BaseHP), MaxBaseHP: maxBaseHP},
			Axis:   {Resources: g.axis.Resources, BaseHP: math.Max(0, g.axis.BaseHP), MaxBaseHP: maxBaseHP},
		},
		Units:    units,
		Trenches: trenches,
	}
}

File: web/game.js
(() => {
  "use strict";

  const lobby = document.querySelector("#lobby");
  const battle = document.querySelector("#battle");
  const soloButton = document.querySelector("#soloButton");
  const createButton = document.querySelector("#createButton");
  const joinForm = document.querySelector("#joinForm");
  const roomCodeInput = document.querySelector("#roomCode");
  const lobbyStatus = document.querySelector("#lobbyStatus");
  const battleStatus = document.querySelector("#battleStatus");
  const roomCodeButton = document.querySelector("#copyCode");
  const waitingCode = document.querySelector("#waitingCode");
  const waitingOverlay = document.querySelector("#waitingOverlay");
  const resultOverlay = document.querySelector("#resultOverlay");
  const resultTitle = document.querySelector("#resultTitle");
  const resultReason = document.querySelector("#resultReason");
  const backToLobby = document.querySelector("#backToLobby");
  const resourcesText = document.querySelector("#resources");
  const phaseText = document.querySelector("#phaseText");
  const timerText = document.querySelector("#timer");
  const sideText = document.querySelector("#sideText");
  const unitButtons = document.querySelector("#unitButtons");
  const canvas = document.querySelector("#gameCanvas");
  const ctx = canvas.getContext("2d");

  const unitOrder = ["rifleman", "assault", "gunner", "tank"];
  const state = {
    socket: null,
    code: "",
    mode: "",
    side: "allies",
    catalog: {},
    match: null,
    displayUnits: new Map(),
    attackPulse: new Map(),
    effects: [],
    connected: false,
  };

  function connect(mode, code = "") {
    closeSocket();
    lobbyStatus.textContent = "Menghubungkan ke medan tempur…";
    const scheme = location.protocol === "https:" ? "wss" : "ws";
    const params = new URLSearchParams({ mode });
    if (code) params.set("code", code.trim().toUpperCase());
    const socket = new WebSocket(`${scheme}://${location.host}/ws?${params}`);
    state.socket = socket;

    socket.addEventListener("open", () => {
      state.connected = true;
      lobbyStatus.textContent = "";
    });

    socket.addEventListener("message", (event) => {
      let message;
      try {
        message = JSON.parse(event.data);
      } catch {
        setBattleStatus("Data server tidak dapat dibaca.");
        return;
      }
      handleMessage(message);
    });

    socket.addEventListener("close", () => {
      state.connected = false;
      if (!battle.classList.contains("hidden") && state.match?.phase !== "finished") {
        setBattleStatus("Koneksi terputus. Kembali ke lobby untuk mencoba lagi.");
      }
    });

    socket.addEventListener("error", () => {
      lobbyStatus.textContent = "Gagal tersambung. Pastikan server berjalan.";
    });
  }

  function handleMessage(message) {
    switch (message.type) {
      case "welcome":
        state.code = message.code;
        state.mode = message.mode;
        state.side = message.side;
        state.catalog = message.catalog || {};
        setupBattle();
        break;
      case "state":
        if (message.state) {
          state.match = message.state;
          syncEffects(message.state.units || []);
          updateHUD();
        }
        break;
      case "notice":
        setBattleStatus(message.message || "");
        window.setTimeout(() => setBattleStatus(""), 2600);
        break;
      case "error":
        if (battle.classList.contains("hidden")) {
          lobbyStatus.textContent = message.message || "Terjadi kesalahan.";
        } else {
          setBattleStatus(message.message || "Perintah ditolak server.");
        }
        break;
      default:
        break;
    }
  }

  function setupBattle() {
    lobby.classList.add("hidden");
    battle.classList.remove("hidden");
    roomCodeButton.textContent = state.mode === "solo" ? "SOLO" : state.code;
    waitingCode.textContent = state.code;
    sideText.textContent = state.side.toUpperCase();
    sideText.style.color = state.side === "allies" ? "#8ab7d7" : "#d78a76";
    buildUnitButtons();
    resizeCanvas();
  }

  function buildUnitButtons() {
    unitButtons.replaceChildren();
    unitOrder.forEach((unitType, index) => {
      const spec = state.catalog[unitType];
      if (!spec) return;
      const button = document.createElement("button");
      button.type = "button";
      button.className = "unit-button";
      button.dataset.unit = unitType;
      button.title = spec.description || spec.name;
      button.innerHTML = `<strong>${index + 1}. ${escapeHTML(spec.name)}</strong><span>${Math.round(spec.cost)} supply</span>`;
      button.addEventListener("click", () => spawn(unitType));
      unitButtons.append(button);
    });
  }

  function spawn(unitType) {
    if (!state.socket || state.socket.readyState !== WebSocket.OPEN || state.match?.phase !== "playing") return;
    state.socket.send(JSON.stringify({ type: "spawn", unit: unitType }));
  }

  function updateHUD() {
    const match = state.match;
    if (!match) return;
    const player = match.players?.[state.side];
    resourcesText.textContent = player ? Math.floor(player.resources) : "0";
    timerText.textContent = formatTime(match.elapsed || 0);
    phaseText.textContent = phaseLabel(match.phase);

    waitingOverlay.classList.toggle("hidden", !(match.phase === "waiting" && state.mode === "pvp"));
    resultOverlay.classList.toggle("hidden", match.phase !== "finished");

    if (match.phase === "finished") {
      const won = match.winner === state.side;
      const draw = match.winner === "draw";
      resultTitle.textContent = draw ? "Seri" : won ? "Kemenangan" : "Kekalahan";
      resultReason.textContent = humanReason(match.reason);
    }

    document.querySelectorAll(".unit-button").forEach((button) => {
      const spec = state.catalog[button.dataset.unit];
      button.disabled = match.phase !== "playing" || !player || player.resources < spec.cost;
    });
  }

  function syncEffects(units) {
    const live = new Set();
    units.forEach((unit) => {
      live.add(unit.id);
      const previousPulse = state.attackPulse.get(unit.id) || 0;
      if (unit.attackPulse > previousPulse) {
        state.effects.push({ x: unit.x, y: unit.y, side: unit.side, life: 0.13, maxLife: 0.13 });
      }
      state.attackPulse.set(unit.id, unit.attackPulse);
      if (!state.displayUnits.has(unit.id)) {
        state.displayUnits.set(unit.id, { x: unit.x, y: unit.y });
      }
    });
    [...state.displayUnits.keys()].forEach((id) => {
      if (!live.has(id)) {
        state.displayUnits.delete(id);
        state.attackPulse.delete(id);
      }
    });
  }

  function renderFrame(timestamp) {
    const dt = Math.min(0.05, (timestamp - (renderFrame.last || timestamp)) / 1000);
    renderFrame.last = timestamp;
    drawGame(dt);
    requestAnimationFrame(renderFrame);
  }

  function drawGame(dt) {
    const width = canvas.clientWidth;
    const height = canvas.clientHeight;
    const dpr = Math.min(window.devicePixelRatio || 1, 2);
    if (canvas.width !== Math.round(width * dpr) || canvas.height !== Math.round(height * dpr)) {
      canvas.width = Math.round(width * dpr);
      canvas.height = Math.round(height * dpr);
    }
    ctx.setTransform(dpr, 0, 0, dpr, 0, 0);

    drawBackground(width, height);
    const match = state.match;
    if (!match) {
      drawCenteredText("Menghubungkan…", width, height);
      return;
    }

    const scaleX = width / match.width;
    const scaleY = height / match.height;
    const groundY = height * 0.61;
    drawTrenches(match.trenches || [], scaleX, groundY, height);
    drawBases(match, scaleX, scaleY, groundY, width);

    const units = [...(match.units || [])].sort((a, b) => a.y - b.y);
    units.forEach((unit) => {
      const visual = state.displayUnits.get(unit.id) || { x: unit.x, y: unit.y };
      const smoothing = 1 - Math.pow(0.0001, dt);
      visual.x += (unit.x - visual.x) * smoothing;
      visual.y += (unit.y - visual.y) * smoothing;
      state.displayUnits.set(unit.id, visual);
      drawUnit(unit, visual.x * scaleX, visual.y * scaleY, scaleX, scaleY);
    });

    state.effects = state.effects.filter((effect) => {
      effect.life -= dt;
      if (effect.life <= 0) return false;
      drawMuzzleFlash(effect, scaleX, scaleY);
      return true;
    });
  }

  function drawBackground(width, height) {
    const sky = ctx.createLinearGradient(0, 0, 0, height);
    sky.addColorStop(0, "#777967");
    sky.addColorStop(0.58, "#a29b79");
    sky.addColorStop(0.59, "#625944");
    sky.addColorStop(1, "#29271e");
    ctx.fillStyle = sky;
    ctx.fillRect(0, 0, width, height);

    ctx.fillStyle = "rgba(48, 47, 39, 0.32)";
    for (let i = 0; i < 9; i += 1) {
      const x = (i * 193 + 70) % width;
      const y = height * (0.16 + (i % 3) * 0.08);
      ctx.beginPath();
      ctx.ellipse(x, y, 70 + (i % 2) * 30, 20, 0, 0, Math.PI * 2);
      ctx.fill();
    }

    ctx.strokeStyle = "rgba(20, 22, 17, 0.35)";
    ctx.lineWidth = 2;
    for (let y = height * 0.66; y < height; y += 28) {
      ctx.beginPath();
      ctx.moveTo(0, y);
      ctx.lineTo(width, y + 5 * Math.sin(y));
      ctx.stroke();
    }
  }

  function drawTrenches(trenches, scaleX, groundY, height) {
    trenches.forEach((trench) => {
      const x = trench.x * scaleX;
      const w = Math.max(18, trench.width * scaleX);
      ctx.fillStyle = "#171812";
      ctx.fillRect(x - w / 2, groundY - 6, w, height - groundY + 6);
      ctx.fillStyle = "#423b2c";
      ctx.fillRect(x - w / 2 - 4, groundY - 10, w + 8, 8);
      ctx.strokeStyle = "rgba(214, 169, 75, 0.22)";
      ctx.setLineDash([5, 7]);
      ctx.strokeRect(x - w / 2, groundY + 10, w, Math.max(30, height - groundY - 28));
      ctx.setLineDash([]);
    });
  }

  function drawBases(match, scaleX, scaleY, groundY, width) {
    drawBase("allies", 72 * scaleX, groundY, match.players.allies, false);
    drawBase("axis", width - 72 * scaleX, groundY, match.players.axis, true);
  }

  function drawBase(side, x, groundY, player, flipped) {
    const color = side === "allies" ? "#557d9b" : "#9c5949";
    ctx.save();
    ctx.translate(x, groundY);
    ctx.scale(flipped ? -1 : 1, 1);
    ctx.fillStyle = "#292a22";
    ctx.fillRect(-32, -88, 64, 88);
    ctx.fillStyle = color;
    ctx.fillRect(-36, -70, 72, 13);
    ctx.fillStyle = "#171812";
    ctx.fillRect(-15, -42, 30, 42);
    ctx.fillStyle = color;
    ctx.beginPath();
    ctx.moveTo(4, -88);
    ctx.lineTo(4, -145);
    ctx.lineTo(52, -126);
    ctx.lineTo(4, -109);
    ctx.closePath();
    ctx.fill();
    ctx.restore();
    drawHealthBar(x - 38, groundY - 165, 76, player.baseHp / player.maxBaseHp, color);
  }

  function drawUnit(unit, x, y, scaleX, scaleY) {
    const spec = state.catalog[unit.type] || {};
    const direction = unit.side === "allies" ? 1 : -1;
    const color = unit.side === "allies" ? "#5b84a6" : "#a75f4e";
    const radius = Math.max(8, (unit.radius || 14) * Math.min(scaleX, scaleY) * 1.25);

    ctx.save();
    ctx.translate(x, y);
    ctx.scale(direction, 1);

    if (unit.type === "tank") {
      ctx.fillStyle = "#2a2d23";
      ctx.fillRect(-radius * 1.2, -radius * 0.25, radius * 2.4, radius * 0.82);
      ctx.strokeStyle = "#11130f";
      ctx.lineWidth = 4;
      ctx.stroke();
      ctx.fillStyle = color;
      ctx.fillRect(-radius * 0.55, -radius * 0.7, radius * 1.2, radius * 0.55);
      ctx.fillStyle = "#1c1e18";
      ctx.fillRect(radius * 0.3, -radius * 0.6, radius * 1.3, radius * 0.12);
      ctx.fillStyle = "#11130f";
      for (let i = -2; i <= 2; i += 1) {
        ctx.beginPath();
        ctx.arc(i * radius * 0.38, radius * 0.42, radius * 0.22, 0, Math.PI * 2);
        ctx.fill();
      }
    } else {
      ctx.strokeStyle = "#1b1c17";
      ctx.lineWidth = Math.max(2, radius * 0.18);
      ctx.beginPath();
      ctx.moveTo(0, radius * 0.1);
      ctx.lineTo(-radius * 0.25, radius * 1.1);
      ctx.moveTo(0, radius * 0.1);
      ctx.lineTo(radius * 0.42, radius * 1.05);
      ctx.stroke();
      ctx.fillStyle = color;
      ctx.fillRect(-radius * 0.55, -radius * 0.55, radius * 1.1, radius * 1.05);
      ctx.fillStyle = "#c7b28b";
      ctx.beginPath();
      ctx.arc(0, -radius * 0.9, radius * 0.43, 0, Math.PI * 2);
      ctx.fill();
      ctx.fillStyle = "#292b22";
      ctx.fillRect(-radius * 0.47, -radius * 1.2, radius * 0.94, radius * 0.28);
      ctx.strokeStyle = "#1b1c17";
      ctx.lineWidth = Math.max(2, radius * 0.18);
      ctx.beginPath();
      ctx.moveTo(radius * 0.2, -radius * 0.3);
      ctx.lineTo(radius * (unit.type === "gunner" ? 1.75 : 1.35), -radius * 0.2);
      ctx.stroke();
      if (unit.type === "assault") {
        ctx.fillStyle = "#d6a94b";
        ctx.fillRect(-radius * 0.18, -radius * 0.28, radius * 0.35, radius * 0.35);
      }
    }
    ctx.restore();

    drawHealthBar(x - radius, y - radius * 1.8, radius * 2, unit.hp / unit.maxHp, color);
    if (spec.name) {
      ctx.fillStyle = "rgba(243, 236, 215, 0.68)";
      ctx.font = "10px system-ui";
      ctx.textAlign = "center";
      ctx.fillText(spec.name, x, y + radius * 1.65);
    }
  }

  function drawHealthBar(x, y, width, ratio, color) {
    const clamped = Math.max(0, Math.min(1, ratio));
    ctx.fillStyle = "rgba(12, 13, 10, 0.72)";
    ctx.fillRect(x, y, width, 5);
    ctx.fillStyle = color;
    ctx.fillRect(x + 1, y + 1, Math.max(0, (width - 2) * clamped), 3);
  }

  function drawMuzzleFlash(effect, scaleX, scaleY) {
    const x = effect.x * scaleX + (effect.side === "allies" ? 18 : -18);
    const y = effect.y * scaleY - 8;
    const alpha = effect.life / effect.maxLife;
    ctx.save();
    ctx.globalAlpha = alpha;
    ctx.fillStyle = "#ffd67a";
    ctx.beginPath();
    ctx.arc(x, y, 4 + (1 - alpha) * 7, 0, Math.PI * 2);
    ctx.fill();
    ctx.restore();
  }

  function drawCenteredText(text, width, height) {
    ctx.fillStyle = "rgba(243, 236, 215, 0.8)";
    ctx.font = "700 20px system-ui";
    ctx.textAlign = "center";
    ctx.fillText(text, width / 2, height / 2);
  }

  function resizeCanvas() {
    const frame = canvas.parentElement;
    const available = Math.max(360, window.innerHeight - 210);
    frame.style.height = `${available}px`;
  }

  function closeSocket() {
    if (state.socket) {
      state.socket.close();
      state.socket = null;
    }
  }

  function returnToLobby() {
    closeSocket();
    state.match = null;
    state.displayUnits.clear();
    state.attackPulse.clear();
    state.effects = [];
    resultOverlay.classList.add("hidden");
    waitingOverlay.classList.add("hidden");
    battle.classList.add("hidden");
    lobby.classList.remove("hidden");
    lobbyStatus.textContent = "";
    battleStatus.textContent = "";
  }

  function setBattleStatus(text) {
    battleStatus.textContent = text;
  }

  function phaseLabel(phase) {
    if (phase === "playing") return "BERTEMPUR";
    if (phase === "finished") return "SELESAI";
    return "MENUNGGU";
  }

  function humanReason(reason) {
    const reasons = {
      "axis headquarters destroyed": "Markas Axis berhasil dihancurkan.",
      "allied headquarters destroyed": "Markas Allies berhasil dihancurkan.",
      "both headquarters were destroyed": "Kedua markas hancur pada saat bersamaan.",
      "opponent disconnected": "Lawan meninggalkan room.",
    };
    return reasons[reason] || reason || "Pertempuran telah berakhir.";
  }

  function formatTime(seconds) {
    const total = Math.max(0, Math.floor(seconds));
    const minutes = String(Math.floor(total / 60)).padStart(2, "0");
    const rest = String(total % 60).padStart(2, "0");
    return `${minutes}:${rest}`;
  }

  function escapeHTML(value) {
    return String(value)
      .replaceAll("&", "&amp;")
      .replaceAll("<", "&lt;")
      .replaceAll(">", "&gt;")
      .replaceAll('"', "&quot;")
      .replaceAll("'", "&#039;");
  }

  soloButton.addEventListener("click", () => connect("solo"));
  createButton.addEventListener("click", () => connect("create"));
  joinForm.addEventListener("submit", (event) => {
    event.preventDefault();
    const code = roomCodeInput.value.trim().toUpperCase();
    if (code.length !== 6) {
      lobbyStatus.textContent = "Kode room harus terdiri dari 6 karakter.";
      return;
    }
    connect("join", code);
  });
  roomCodeInput.addEventListener("input", () => {
    roomCodeInput.value = roomCodeInput.value.toUpperCase().replace(/[^A-Z0-9]/g, "").slice(0, 6);
  });
  roomCodeButton.addEventListener("click", async () => {
    if (state.mode !== "pvp" || !state.code) return;
    try {
      await navigator.clipboard.writeText(state.code);
      setBattleStatus("Kode room disalin.");
    } catch {
      setBattleStatus(`Kode room: ${state.code}`);
    }
  });
  backToLobby.addEventListener("click", returnToLobby);
  window.addEventListener("resize", resizeCanvas);
  window.addEventListener("keydown", (event) => {
    if (event.repeat || battle.classList.contains("hidden")) return;
    const index = Number(event.key) - 1;
    if (index >= 0 && index < unitOrder.length) spawn(unitOrder[index]);
  });

  requestAnimationFrame(renderFrame);
})();
